### PR TITLE
Use version 2 STS tokens to query opt-in regions

### PR DESCRIPTION
**Pull request by iainelder**
* Refactor client and session init to enable further improvements
* Add tests for assuming_session
* Use assuming_session tests to fix a bug caused by setting None as the region
* Explicitly suppress the region output for a single-region query for backwards compatibility
* Set the default region in other tests to make them pass again (Not 100% sure why it's needed now. Something to do with moto's lax validation? I think changing the STS client behavior makes it necessary)
* Delete broken tests that don't use public interfaces

The new tests use the `cove()` form because the tests are easier to read and write. I trust that Python's syntactic sugar makes it work the same way for the `@cove` form.

The only thing that makes me uneasy about this PR is not having a clear answer for why I have to set the default region.

If you want I can split this into two PRs; one to fix the bug in `assuming_session` and another to add support to query opt-in regions.

There are no unit tests for opt-in region support because Moto doesn't implement the concept. Use the test in #77 to check it.

### CODE REVIEW

File: tests/test_regions.py [hunk 1]
```diff
@@ -1,81 +1,70 @@
+import functools
+
 import pytest
+from _pytest.monkeypatch import MonkeyPatch
 from boto3.session import Session
 
 from botocove import cove
 from tests.moto_mock_org.moto_models import SmallOrg
 
+cove = functools.partial(cove, thread_workers=1, raise_exception=True)
```
Comment (iainelder) [inline tests/test_regions.py:10]: In real use cases I almost always use the default settings for `thread_workers` and `raise_exception`.

But setting them like this makes the tests easier to read, write and debug.

* We can assume that `ExceptionDetails` and `FailedAssumeRole` are empty and just check `Results`.
* We can use `pytest.raises` to catch any exception.
* We can set a breakpoint in threaded code and get just one debugger instance.

I'd like to refactor all the tests to use this setup.

I'd rewrite `test_exceptions` to check that `ExceptionDetails` and `FailedAssumeRole` are set correctly for a few examples when `raise_exception` is `False`.

I would also add a test for the threading.

File: botocove/cove_decorator.py [hunk 2]
```diff
@@ -64,26 +65,35 @@ def wrapper(*args: Any, **kwargs: Any) -> CoveOutput:
...
                 ],
                 FailedAssumeRole=[
                     {k: v for k, v in f.items() if v is not None}
-                    for f in output["Exceptions"]
+                    for f in function_output["Exceptions"]
                     if f["AssumeRoleSuccess"] is False
                 ],
             )
 
+            if regions is None:
```
Comment (connelldave) [inline botocove/cove_decorator.py:87]: I'm not following why this is required - I'm sure I'm missing something obvious but could you explain?

**Review by connelldave — COMMENTED** [review]
This feels like three PRs worth of changes wrapped into one - the cove_host_account.py changes look like a tidy refactor of clunky code, I haven't tried to read the test diffs yet, and I don't understand the addition of deletion of a region key :)

**Comment by iainelder** [comment]
"TL;DR" is good feedback for this. I was thinking the same.

I'll break it into the three parts you identified over the next few days so it's easier to reason about. First the refactoring, then the `assuming_session` bugfix, then finally the version 2 STS support that keeps botocove relevant in new AWS regions.

File: botocove/cove_decorator.py (hunk 2 above)
Comment (connelldave) [reply botocove/cove_decorator.py:87]: Gotcha - makes sense. I don't think we need to maintain this behaviour in a new release: we can just always emit a Region key in the output and bump the sem ver appropriately. I'm not overly worried about adding to output, it's removing things that's more of a place to be wary of previous API contracts.

**Comment by connelldave** [comment]
I wouldn't go as far as to be ungrateful and say TLDR :) It's just that when all the tests change, and then there's new logic that doesn't appear to impact something in the diff (the del regions loop), plus positive refactors of old behaviour, it's hard to reason about without relearning the whole code base.

**Comment by iainelder** [comment]
I know, I didn't take it badly :-) I've already confused myself about why exactly some of the test changes were necessary, so I don't want to make you spend time trying to figure it out. I'll break it up into smaller PRs that are easier to digest. Probably won't get time this week, but I hope to submit the first one next week.

**Pull request closed without merging**